runtime/src/main/java/org/corfudb/runtime/view/stream/AddressMapStreamView.java

```diff
@@ -155,53 +155,74 @@ protected boolean discoverAddressSpace(final UUID streamId,
-                                    "the checkpoint capability.", this, trimMark, stopAddress);
-                    log.info(message);
-                    throw new TrimmedException(message);
-                }
-                // Address maps might have been trimmed, hence not reflecting all updates to the stream
-                // For this reason, in the case of a valid trim mark, we must be sure this space is
-                // already resolved or loaded by a checkpoint.
-                if (isCheckpointCapable()
-                        && Address.isAddress(trimMark)
-                        && !isTrimCoveredByCheckpointOrLocalView(trimMark)) {
-                    if (getReadOptions().isIgnoreTrim()) {
-                        log.debug("getStreamAddressMap[{}]: Ignoring trimmed exception for address[{}].",
-                                this, streamAddressSpace.getTrimMark());
-                    } else {
-                        String message = String.format("getStreamAddressMap[{%s}] [%d, %d] " +
-                                        "stream has been trimmed at address %s and this space is " +
-                                        "not covered by the loaded checkpoint with start " +
-                                        "address %s, while accessing the stream at version " +
-                                        "%s. Looking for a new checkpoint.",
-                                this, stopAddress, startAddress, trimMark,
-                                getCurrentContext().getCheckpoint().startAddress, maxGlobal);
-                        throw new TrimmedException(message);
-                    }
-                }
+                moveToReadQueue(streamAddressSpace, queue, startAddress, stopAddress, maxGlobal);
             }
         }
 
         addressCount += queue.size();
         return !queue.isEmpty();
     }
 
+    private void moveToReadQueue(final StreamAddressSpace streamAddressSpace,
+                                  final NavigableSet<Long> queue,
+                                  final long startAddress,
+                                  final long stopAddress,
+                                  final long maxGlobal) {
+
+        // Transfer discovered addresses to queue. We must limit to maxGlobal,
+        // as startAddress could be ahead of maxGlobal---in case it reflects
+        // the tail of the stream.
+        queue.addAll(streamAddressSpace.copyAddressesToSet(maxGlobal));
+
+        final long trimMark = streamAddressSpace.getTrimMark();
+
+        // No valid trim-mark has been defined,
+        // thus we cannot run into trimmed address space.
+        if (!Address.isAddress(trimMark)) {
+            return;
+        }
+
+        // Trim-mark is below or equal to the lowest address,
+        // thus we cannot run into trimmed address space.
+        if (trimMark <= stopAddress) {
+            return;
+        }
```

Comment:
I think we can combine these two conditions.

Reply:
I decided to make them separate for readability purposes.